app/lib/synthetic_study_populator.rb

```diff
@@ -77,67 +80,97 @@ def self.add_files(study, study_config, synthetic_study_folder, user)
-        end
+  def self.process_expression_file_params(study, file_info)
+    exp_params = {}
+    if file_info['type'] == 'Expression Matrix'
+      exp_finfo_params = file_info['expression_file_info']
+      if exp_finfo_params.present?
+        exp_file_info = ExpressionFileInfo.new(
+          is_raw_counts: exp_finfo_params['is_raw_counts'] ? true : false,
+          units: exp_finfo_params['units'],
+          biosample_input_type: exp_finfo_params['biosample_input_type'],
+          library_preparation_protocol: exp_finfo_params['library_preparation_protocol'],
+          modality: exp_finfo_params['modality']
+        )
+        exp_params['expression_file_info'] = exp_file_info
       end
+    end
+    exp_params
+  end
 
-      study_file = StudyFile.create!(study_file_params)
-      # the status has to be 'uploading' when created so the file gets pulled into the workspace
-      # after creation, we want it to be uploaded so that, e.g. bundles can create
-      study_file.update(status: 'uploaded')
+  def self.process_label_file_params(study, file_info)
+    params = {}
+    if file_info['type'] == 'Coordinate Labels'
+      if !file_info['cluster_file_name']
+        throw 'Coordinate label files must specify a cluster_file_name'
+      end
+      matching_cluster_file = StudyFile.find_by(name: file_info['cluster_file_name'], study: study)
+      if matching_cluster_file.nil?
+        throw "No cluster file with name #{file_info['cluster_file_name']} to match coordinate labels"
+      end
+      params[:options] = {'cluster_file_id' => matching_cluster_file.id.to_s}
+    end
+    params
+  end
 
-      if !study.detached
-        FileParseService.run_parse_job(study_file, study, user)
+  def self.process_sequence_file_params(study, file_info)
+    params = {}
+    if file_info['type'] == 'BAM Index'
+      if !file_info['bam_file_name']
+        throw 'BAM index files must specify a bam_file_name'
+      end
+      matching_bam_file = StudyFile.find_by(name: file_info['bam_file_name'], study: study)
+      if matching_bam_file.nil?
+        throw "No BAM file with name #{file_info['bam_file_name']} to match bai file"
       end
+      params[:options] = {'bam_id' => matching_bam_file.id.to_s}
     end
+    params
   end
 
+
   # process coordinate/cluster arguments, return a hash of params suitable for passing to a StudyFile constructor
-  def self.process_coordinate_file_params(file_info)
+  def self.process_coordinate_file_params(study, file_info)
     params = {}
     if !file_info['is_spatial'].nil?
       params[:is_spatial] = file_info['is_spatial']
     end
+    if file_info['spatial_cluster_associations'].present?
+      # look up the ids for the associations
+      # note that this requires the associated file to have already been added
+      params[:spatial_cluster_associations] = file_info['spatial_cluster_associations'].map do |cluster_file_name|
+        StudyFile.find_by!(study: study, name: cluster_file_name).id.to_s
```

Comment:
This would fail gracefully and therefore avoid the issue of the file not being present yet
```suggestion
        StudyFile.find_by!(study: study, name: cluster_file_name).try(:id).to_s
```

Reply:
I actually want this to fail hard -- the particular way the populator assembles bundle params requires the cluster file to be defined before the spatial one is.

Reply:
Fair point.